src/Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/SetPiFromGauge.hpp

```diff
@@ -0,0 +1,69 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <memory>
+#include <random>
+#include <string>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Domain/ElementMap.hpp"
+#include "Domain/FunctionsOfTime/FunctionOfTime.hpp"
+#include "Domain/FunctionsOfTime/Tags.hpp"
+#include "Domain/Tags.hpp"
+#include "Domain/TagsTimeDependent.hpp"
+#include "Evolution/Initialization/Tags.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/GaugeSourceFunctions/Tags/GaugeCondition.hpp"
+#include "Evolution/Systems/GeneralizedHarmonic/Tags.hpp"
+#include "Time/Tags.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace GeneralizedHarmonic::gauges {
+/*!
+ * \brief Set \f$\Pi_{ab}\f$ from the gauge source function.
+ *
+ * This is necessary to ensure the initial data is in the desired evolution
+ * gauge.
+ */
+template <size_t Dim>
+struct SetPiFromGauge {
+ public:
+  using return_tags =
+      tmpl::list<GeneralizedHarmonic::Tags::Pi<Dim, Frame::Inertial>>;
+  using argument_tags =
+      tmpl::list<::Tags::Time, domain::Tags::Mesh<Dim>,
+                 domain::Tags::ElementMap<Dim, Frame::Grid>,
+                 domain::CoordinateMaps::Tags::CoordinateMap<Dim, Frame::Grid,
+                                                             Frame::Inertial>,
+                 domain::Tags::FunctionsOfTime,
+                 domain::Tags::Coordinates<Dim, Frame::ElementLogical>,
```

Comment:
I'm a little confused now. If things are valid during initialization, why not just only add `domain::Tags::InverseJacobian<Dim, Frame::ElementLogical, Frame::Inertial>` here so the compute tag does the work, and you don't have to calculate the coordinates explicitly yourself? If things aren't valid during initialization, could you explain why?

Reply:
I don't know if they are valid now. They might be because of the initialization tag change Larry did, but there are several other places that compute it this way too, and I think we should change them all at once if we do change it.

Reply:
This depends upon whether you intend this function to be applicable for dg only, or for subcell as well.  If dg only, then you can just grab the appropriate compute tags.  If you want this to work for both dg and subcell, then you need to template on the types of any tags that are different for dg/subcell (currently the mesh and logical coordinates). (This is why I think SetVariables is templated on LogicalCoordinatesTag)

Reply:
All of GH needs fairly extensive subcell support...